flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestServerEndpoint.java

```diff
@@ -0,0 +1,172 @@
+	 */
+	protected abstract Collection<AbstractRestHandler<?, ?>> initializeHandlers();
+
+	/**
+	 * Starts this REST server endpoint.
+	 */
+	public void start() {
+		log.info("Starting rest endpoint.");
+		initializeHandlers()
+			.forEach(this::registerHandler);
+
+		ChannelInitializer<SocketChannel> initializer = new ChannelInitializer<SocketChannel>() {
+
+			@Override
+			protected void initChannel(SocketChannel ch) {
+				Handler handler = new RouterHandler(router);
+
+				// SSL should be the first handler in the pipeline
+				if (sslEngine != null) {
+					ch.pipeline().addLast("ssl", new SslHandler(sslEngine));
+				}
+
+				ch.pipeline()
+					.addLast(new HttpServerCodec())
+					.addLast(new HttpObjectAggregator(1024 * 1024 * 10))
+					.addLast(handler.name(), handler)
+					.addLast(new PipelineErrorHandler(log));
+			}
+		};
+
+		NioEventLoopGroup bossGroup = new NioEventLoopGroup(1);
+		NioEventLoopGroup workerGroup = new NioEventLoopGroup();
+
+		bootstrap = new ServerBootstrap();
+		bootstrap
+			.group(bossGroup, workerGroup)
+			.channel(NioServerSocketChannel.class)
+			.childHandler(initializer);
+
+		ChannelFuture ch;
+		if (configuredAddress == null) {
+			ch = bootstrap.bind(configuredPort);
+		} else {
+			ch = bootstrap.bind(configuredAddress, configuredPort);
+		}
+		serverChannel = ch.syncUninterruptibly().channel();
+
+		InetSocketAddress bindAddress = (InetSocketAddress) serverChannel.localAddress();
+		String address = bindAddress.getAddress().getHostAddress();
+		int port = bindAddress.getPort();
+
+		log.info("Rest endpoint listening at {}" + ':' + "{}", address, port);
+	}
+
+	private <R extends RequestBody, P extends ResponseBody> void registerHandler(AbstractRestHandler<R, P> handler) {
+		switch (handler.getMessageHeaders().getHttpMethod()) {
+			case GET:
+				router.GET(handler.getMessageHeaders().getTargetRestEndpointURL(), handler);
+				break;
+			case POST:
+				router.POST(handler.getMessageHeaders().getTargetRestEndpointURL(), handler);
+				break;
+		}
+	}
+
+	/**
+	 * Returns the address on which this endpoint is accepting requests.
+	 *
+	 * @return address on which this endpoint is accepting requests
+	 */
+	public InetSocketAddress getServerAddress() {
+		Channel server = this.serverChannel;
+		if (server != null) {
+			try {
+				return ((InetSocketAddress) server.localAddress());
+			} catch (Exception e) {
+				log.error("Cannot access local server address", e);
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Stops this REST server endpoint.
+	 */
+	public void shutdown() {
+		log.info("Shutting down rest endpoint.");
+		if (this.serverChannel != null) {
+			this.serverChannel.close().awaitUninterruptibly();
+		}
+		if (bootstrap != null) {
+			if (bootstrap.group() != null) {
+				bootstrap.group().shutdownGracefully();
+			}
+			if (bootstrap.childGroup() != null) {
+				bootstrap.childGroup().shutdownGracefully();
+			}
+		}
+		log.info("Rest endpoint shutdown complete.");
```

Comment:
I think in order to say that we have to collect the termination futures of the groups and then wait on their completion. This should also have a timeout.